lib/oplog/tail.go

```diff
@@ -372,11 +366,42 @@ func closeCursor(cursor *mongo.Cursor) {
 // ignored it or failed at some later step.
 func (tailer *Tailer) unmarshalEntry(rawData bson.Raw, denylist *sync.Map, readOrdinal int) (timestamp *primitive.Timestamp, pubs []*redispub.Publication, sendMetricsData func()) {
 	var result rawOplogEntry
+	var ok bool
+	result.Namespace, ok = rawData.Lookup("ns").StringValueOK()
+	if !ok {
+		log.Log.Error("Error unmarshalling oplog namespace entry")
+	}
 
-	err := bson.Unmarshal(rawData, &result)
-	if err != nil {
-		log.Log.Errorw("Error unmarshalling oplog entry", "error", err)
-		return
+	if len(result.Namespace) > 0 { // try to filter early if possible
+		db, _ := parseNamespace(result.Namespace)
+
+		if _, denied := denylist.Load(db); denied {
+			log.Log.Debugw("Skipping oplog entry", "database", db)
+			metricOplogEntriesFiltered.WithLabelValues(db).Add(1)
+
+			return
+		}
+	}
+
+	t, i, ok := rawData.Lookup("ts").TimestampOK()
+	if !ok {
+		log.Log.Warn("Error unmarshalling oplog timestamp entry")
+	}
+	result.Timestamp = primitive.Timestamp{T: t, I: i}
+
+	result.Operation, ok = rawData.Lookup("op").StringValueOK()
+	if !ok {
+		log.Log.Warn("Error unmarshalling oplog operation entry")
+	}
+
+	result.Doc, ok = rawData.Lookup("o").DocumentOK()
+	if !ok {
+		log.Log.Warn("Error unmarshalling oplog document entry")
+	}
+
+	result.Update, ok = rawData.Lookup("o2").DocumentOK()
+	if !ok {
+		log.Log.Debug("Error unmarshalling oplog update entry")
 	}
```

Comment:
Parsing the namespace with a manual lookup to determine if we discard the OpLog makes sense, but can't the rest of this just use the `Unmarshal` code from before?

Reply:
It could, and it would be simpler, but there's still a lot of gain to be had by using bson all the way.  The unmarshal seems to need to allocate more or something in a way that increases resource usage as the oplog entry size increases.  With the bson methods, the time and memory remains more or less the same (and significantly better), regardless of oplog size.

Reply:
That is disappointing. It seems like all of our current issues in OTR and IM are due to reflection-based deserialization 😞